// node_bis/index.js
var fs          = require("fs");
var http        = require("http");
var request 		= require('request');
var qs 					= require('querystring');
var multiparter = require("multiparter");

var nodeBIS = function(key, url, path){

	function addImage(localPath, file, remoteFile, storagePath, callback){

		var err = null;
		var request = new multiparter.request(http, {
			host: this.url,
			port: 80,
			path: this.path + 'api.php',
			method: 'POST'
		});

		// add some plain params here
		request.setParam("cmd", "addImageByForm");
		request.setParam("key", this.key);
		request.setParam("storage_id", "2");
		request.setParam("imagePath", storagePath);		

		request.addStream(
				"filename[]",
				remoteFile,
				"image/gif",
				fs.statSync(localPath + file).size,
				fs.createReadStream(localPath + file)
		);

		// send request and receive response
		request.send(function(error, response) {

			if (error) {
				return(callback(false, error));
			}
	
			var data = "";
	
			response.setEncoding("utf8");
	
			response.on("data", function(chunk) {
					data += chunk;
			});
	
			response.on("end", function() {
				return(callback(data, err));
			});
	
			response.on("error", function(error) {
					return(callback(false, error));
			});
		});

	}

	function getImages( params, callback ) { 
        params.cmd = 'imageList';
		request.get({
			url: 'http://' + this.url + this.path + 'api.php?' + qs.stringify( params )
		}, function( error, response, data ) {
			if ( error ) {
				return callback( false, error );
			} else {
				data = JSON.parse( data );
				if ( data.success ) {
					return callback( data.records );
				} else {
					return callback( false, data.error );
				}
			}
		});
	}

	function getImageInfo(params, callback) {
		var params = {
				cmd: "getImageInfo"
			,	barcode: params.barcode || ''
			,	image_id: params.image_id || ''
			,	size: params.size || ''
		}

		request.get({
			url: "http://" + this.url + this.path + 'api.php?' + qs.stringify(params)
		}, function(error, response, data) {		
			if (error) {
				return(callback(false));
			} else {
				data = JSON.parse(data);
				if (data.success) {
					return(callback(data.data));
				} else {
					return(callback(false, data.error));
				}
			}
		});
	}

	function getImageUrl(params, callback) {
		var params = {
				cmd: "getImageUrl"
			,	barcode: params.barcode || ''
			,	image_id: params.image_id || ''
			,	size: params.size || ''
		}

		request.get({
			url: "http://" + this.url + this.path + 'api.php?' + qs.stringify(params)
		}, function(error, response, data) {		
			if (error) {
				return(callback(false));
			} else {
				if (data != '' && data[0] == '{') {
					data = JSON.parse(data);
					return(callback(false, data.error));
				} else {
					return(callback(data));
				}
			}
		});
	}

	function listStorage(callback) {
		var params = {
			cmd: "listStorageDevices"
		}

		request.get({
			url: "http://" + this.url + this.path + 'api.php?' + qs.stringify(params)
		}, function(error, response, data) {		
			if (error) {
				return(callback(false));
			} else {
				data = JSON.parse(data);
				if (data.success) {
					return(callback(data.data));
				} else {
					return(callback(false, data.error));
				}
			}
		});
	}

	function sendPackage( batch, callback ) {
		request.post( 'http://' + this.url + this.path + 'api.php', { form: { cmd: 'package', authMode: 'key', key: this.key, data: JSON.stringify(batch) } }, function( error, response, data ) {
            var res = JSON.parse( data );
			if ( error ) {
				return callback( false, error );
			} else {
				if ( res.success ) {
					return callback( res );
				} else {
					return callback( false, res.error );
				}
			}
		});
	}
	
	return {
		addImage: addImage,
		listStorage: listStorage,
		getImages: getImages,
		getImageUrl: getImageUrl,
		getImageInfo: getImageInfo,
		sendPackage: sendPackage,
		key: key || '',
		url: url || '',
		path: path || '',
	};

}

module.exports = nodeBIS;
