// node_bis/index.js
var fs          = require("fs");
var http        = require("http");
var request 		= require('request');
var qs 					= require('querystring');
//var multiparter = require("multiparter");

var nodeBIS = function(key, url, path){

	function addImage(file, callback){
/*		
		var err = null;
		var request = new multiparter.request(http, {
			host: this.url,
			port: 80,
			path: this.path + 'api.php',
			method: 'POST'
		});

		// add some plain params here
		request.setParam("cmd", "createObject");
		request.setParam("key", this.key);
		
		request.addStream(
				"filename",
				"test",
				"image/gif",
				fs.statSync(file).size,
				fs.createReadStream(file)
		);

		// send request and receive response
		request.send(function(error, response) {

			if (error) {
				return(callback(error, null));
			}
	
			var data = "";
	
			response.setEncoding("utf8");
	
			response.on("data", function(chunk) {
					data += chunk;
			});
	
			response.on("end", function() {
				return(callback(err, data));
			});
	
			response.on("error", function(error) {
					return(callback(error, null));
			});
		});
*/
	}
	
	function listStorage(callback) {
		var params = {
			cmd: "listStorageDevices"
		}

		request.get({
			url: this.url + 'api.php?' + qs.stringify(params)
		}, function(error, response, data) {
			if (error) {
				return(callback(false));
			} else {
				data = JSON.parse(data);
				if (data.success) {
					return(callback(data.data));
				}

			}
		});
	}
	
	return {
		addImage: addImage,
		listStorage: listStorage,
		key: key || '',
		url: url || '',
	};

}

module.exports = nodeBIS;
