// bis/index.js

var	path = require('path');
var _ = require('underscore');
var sprintf = require('sprintf').sprintf;
var fs = require('fs');
var mkdirp = require('mkdirp');
var	gm = require('gm');
var mysql = require('mysql');

var BIS = function(params){
	var connection = mysql.createConnection({
	  host     : 'localhost',
	  user     : 'root',
	  password : '',
	  database : 'bis',
	});
	connection.connect();

	var loopFlag = false;
	
	// Image Functions
	
	function loadImageById(imageId,callback) {
		imageId = ('undefined' != typeof imageId && imageId != '') ? imageId : '';
		if(imageId = '') {
			return callback(false);
		} else {
			connection.query(' SELECT * FROM image WHERE image_id = ? ', [imageId], function(err, rows) {
				if(!err) {
					return (rows.length <= 0) ? callback(false) : callback(rows[0]);
				} else {
					return callback(false);
				}
			});
		}
	}
	
	// Image Manipulation
	
	function createThumb(params,callback) {
		params = ('undefined' != typeof params) ? params : '';
		params.imagePath = ('undefined' != typeof params.imagePath && params.imagePath != '') ? params.imagePath : '';
		params.width = ('undefined' != typeof params.width && params.width != '') ? params.width : '';
		params.height = ('undefined' != typeof params.height && params.height != '') ? params.height : '';
		params.postFix = ('undefined' != typeof params.postFix && params.postFix != '') ? params.postFix : '';
		if(params.imagePath == '' || params.width == '' || params.height) {
			return callback(false);
		} else {
			if(path.existsSync(params.imagePath)) {
				var destination = path.dirname(params.imagePath) + '\\' + path.basename(params.imagePath,path.extname(params.imagePath)) + params.postFix + path.extname(params.imagePath);
				gm( params.imagePath ).thumb(params.width, params.height, destination, 90 , function(err) {
					return callback(true);
				});
			} else {
				return callback(false);
			}
		}
	}
	
	// Process Queue Functions
	
	function getPQueueCount(params,callback) {
		params = ('undefined' != typeof params) ? params : '';
		params.processType = ('undefined' != typeof params.processType && params.processType !='' && -1 != _.indexOf(['flickr_add','picassa_add','zoomify','google_tile','ocr_add','box_add','name_add','evernote','all'],params.processType)) ? params.processType : '';
		var query = ' SELECT count(*) ct FROM process_queue ';
		if(params.processType != '') {
			query += sprintf(" WHERE process_type = '%s' ", connection.escape(params.processType));
		}
		connection.query(query, function(err, rows, fields) {
			if (!err) {
				console.log(rows);
				return callback(rows[0].ct);
			} else {
				console.log(err);
				return callback(false);
			}
		});
		
	}
	
	function deleteProcessQueue(params,calback) {
		params = ('undefined' != typeof params) ? params : '';
		params.imageId = ('undefined' != typeof params.imageId && params.imageId != '') ? params.imageId : '';
		params.processType = ('undefined' != typeof params.processType && params.processType != '') ? params.processType : '';
		if(params.imageId == '' || params.processType == '') {
			return callback(false);
		} else {
			connection.query('DELETE FROM process_queue WHERE image_id = ? AND process_type = ? ',[params.imageId,params.processType],function(err, result){
				return (!err) ? callback(true) : callback(false);
			});
		}
		
	}
	
	function popQueue(processType,callback) {
		processType = ('undefined' != typeof processType && processType != '') ? processType : '';
		var query;
		var me = this;
		switch(processType) {
			case 'flickr_add':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'flickr_add' ORDER BY `date_added` LIMIT 1";
				break;
			case 'picassa_add':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'picassa_add' ORDER BY `date_added` LIMIT 1";
				break;
			case 'zoomify':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'zoomify' ORDER BY `date_added` LIMIT 1";
				break;
			case 'google_tile':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'google_tile' ORDER BY `date_added` LIMIT 1";
				break;
			case 'ocr_add':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'ocr_add' ORDER BY `date_added` LIMIT 1";
				break;
			case 'box_add':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'box_add' ORDER BY `date_added` LIMIT 1";
				break;
			case 'name_add':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'name_add' ORDER BY `date_added` LIMIT 1";
				break;
			case 'evernote':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'evernote' ORDER BY `date_added` LIMIT 1";
				break;
			case 'all':
				query = "SELECT * FROM `process_queue` WHERE `process_type` = 'all' ORDER BY `date_added` LIMIT 1";
				break;
			default:
				query = "SELECT * FROM `process_queue` WHERE `process_type` NOT IN ('picassa_add','flickr_add','ocr_add','box_add') ORDER BY `date_added` LIMIT 1";
				break;
		}
		
		connection.query(query, function(err, rows, fields) {
			if (!err) {
				me.deleteProcessQueue({imageId : rows[0].image_id, processType : rows[0].process_type},function(status){
					return (status) ? callback(rows[0]) : callback(false) ;
				});
			} else {
				console.log(err);
				return callback(false);
			}
		});
	}
	
	function processType(record,callback) {
		var me = this;
		record = ('undefined' != typeof record && record != '') ? record : '';
		record.image_id = ('undefined' != typeof record.image_id && record.image_id != '') ? record.image_id : '';
		this.loadImageById(record.image_id,function(imageRow){
			if(false == imageRow) {
				return callback(false);
			} else {
				// TODO integrate storage class code (currently included local paths only)
				
				var image = imageRow.filename;
				var imagePath = imageRow.path;
				switch(imageRow.process_type) {
					case 'all':
						me.createThumb({},function() {
						
						});
						break;
				}

			}
		});
	}
	
	function processQueue(params) {
		var me = this;
		params = ('undefined' != typeof params) ? params : '';
		params.limit = ('undefined' != typeof params.limit && params.limit !='') ? params.limit : '';
		loopFlag = true;
		var processFn = function() {
			if(loopFlag) {
				me.popQueue(function(record){
					if(false != record) {
						me.processType(record,function(){
						
						});
					}
				});
				// pocessFn();
			}
		};
		processFn();

	}

	function barcodePath( barcode ) {
		var id = barcode,loop_flag = true,i = 0,prefix = '',destPath = '';
						
		if (id.length > 8){
			while(loop_flag){
				if(barcode.substr(i) * 1) {
					loop_flag = false;
				} else {
					i++;
				}
				if (i>8) loop_flag = false;
			}
			prefix = id.substr(0,i);
			if (id.substr(i, 1) == "-") {
				prefix += "-";
			}
			id = Math.abs(id.substr(i));
		} else {		
			prefix = "";
		}
		destPath = prefix + "/";
		destPath += parseInt(Math.abs(id / 1000000) + '') + "/";
		destPath += parseInt(Math.abs( (id % 1000000) / 10000) + '') + "/";
		destPath += parseInt(Math.abs( (id % 10000) / 100) + '') + "/";
		destPath += parseInt(Math.abs( (id % 100)) + '') + "/";
		return( destPath );
	}
	
	return {
		barcodePath 	: barcodePath,
		getPQueueCount 	: getPQueueCount,
		processQueue 	: processQueue,
		popQueue	 	: popQueue,
		deleteProcessQueue : deleteProcessQueue,
		processType : processType,
		loadImageById : loadImageById,
		createThumb : createThumb
	};

}

module.exports = BIS;
